### PR TITLE
Pagination and Filtering

### PR DESCRIPTION
**Pull request by Phoeneix**
Github Issue: #100

### Description

Add test cases for
- pagination
- filter

### Overview of work done

    - Added pagination for method GetProductsOfCurrentUser, GetStatusByPrevious
    - Added filtering for GetResterDefintionsOfCurrentUser
    - Created Product and Status API test cases for pagination check
    - Created RasterDefinition API test for filtering check
    -
### Overview of verification done

Executed on UAT

### Overview of integration done

_Explain how this change was integration tested. Provide screenshots or logs if appropriate. An example of this would be a local Harmony deployment._

## PR checklist:

* [ ] Linted
* [ ] Updated unit tests
* [x] Updated changelog
* [ ] Integration testing

_See [Pull Request Review Checklist](../CONTRIBUTING.md#reviewing) for pointers on reviewing this pull request_

### CODE REVIEW

File: API/SWODLR/rasterdefinition.py [hunk 1]
```diff
@@ -13,17 +13,48 @@
 class RasterDefinition():
     
     def GetResterDefintionsOfCurrentUser(
```
Comment (joshgarde) [inline API/SWODLR/rasterdefinition.py]: Misspelling of `Raster`?

File: API/SWODLR/product.py [hunk 2]
```diff
@@ -23,7 +25,10 @@ def GetProductsOfCurrentUser(
         graphQlBody = '''
             {
             	currentUser {
-            		products {
+            		products(limit: ''' + str(pageSize)
+        if pageAfterId != "":
+            graphQlBody += f', after: "{pageAfterId}"'
+        graphQlBody += ''') {
```
Comment (joshgarde) [inline API/SWODLR/product.py]: Is there a way to clean this up?

File: API/SWODLR/rasterdefinition.py [hunk 3]
```diff
@@ -13,17 +13,48 @@
...
+            graphQlBody += f'outputSamplingGridType: "{outputSamplingGridType}"'
+        if rasterResolution != 0:
+            graphQlBody += f'rasterResolution: {rasterResolution}'
+        if utmZoneAdjust != 0:
+            graphQlBody += f'utmZoneAdjust: {utmZoneAdjust}'
+        if mgrsBandAdjust != 0:
+            graphQlBody += f'mgrsBandAdjust: {mgrsBandAdjust}'
+        if filteringPresent:
+            graphQlBody += ')'
+        graphQlBody += ''' {
```
Comment (joshgarde) [inline API/SWODLR/rasterdefinition.py]: Is there a way to clean this up for readability?

File: tests/test_product.py [hunk 4]
```diff
@@ -446,3 +446,28 @@ def test_Product_GetProductsOfCurrentUser_InvalidAuth_401(self):
...
+            if len(products) == pageSize:                
+                lastId = products[count-1]['id']
+                print(f'LastId: {lastId}')
+                response:Response = Product.GetProductsOfCurrentUser(pageSize = pageSize, pageAfterId = lastId)
+                assert response.status_code == expectedStatusCode, f'Response code "{response.status_code}" is not "{expectedStatusCode}"!'
+                page += 1
+            else:
+                count = 0
+        assert page > 0, f'No pagination happened!' 
+
```
Comment (joshgarde) [inline tests/test_product.py:473]: Newline

File: tests/test_rasterdefinition.py [hunk 5]
```diff
@@ -379,3 +413,47 @@ def test_Rasterdefinition_DeleteRasterDefinition_Type_Wrong_RasterDefinitionID(s
             expectedType = expectedType,
             checkForData = False
         )
+
+
+# ========================================== Private Methods ==========================================
```
Comment (joshgarde) [inline tests/test_rasterdefinition.py:418]: Shouldn't private methods be marked with standard PEP8 format?

https://peps.python.org/pep-0008/#descriptive-naming-styles

**Review by joshgarde — CHANGES_REQUESTED** [review]

File: API/SWODLR/product.py (hunk 2 above)
Comment (Phoeneix) [reply API/SWODLR/product.py]: I am not aware of of any way for that. GraphQL requests a string and I can't leave parameters empty, so depending on the inputs the string need to be put together to reflect that.

File: tests/test_rasterdefinition.py (hunk 5 above)
Comment (Phoeneix) [reply tests/test_rasterdefinition.py:418]: I think my naming convention fitting into what's written there:
There should be no difference between method names, as there are no truly private methods in python. Not like in C# or Java.
[Class Names](https://peps.python.org/pep-0008/#class-names)
Class names should normally use the CapWords convention.

The naming convention for functions may be used instead in cases where the interface is documented and used primarily as a callable.

Note that there is a separate convention for builtin names: most builtin names are single words (or two words run together), with the CapWords convention used only for exception names and builtin constants.

File: API/SWODLR/rasterdefinition.py (hunk 1 above)
Comment (Phoeneix) [reply API/SWODLR/rasterdefinition.py]: Fixed

File: API/SWODLR/rasterdefinition.py (hunk 3 above)
Comment (Phoeneix) [reply API/SWODLR/rasterdefinition.py]: Edited:
Removed

File: API/SWODLR/product.py (hunk 2 above)
Comment (joshgarde) [reply API/SWODLR/product.py]: GraphQL does provide a facility to utilize placeholder variables within a query string and replace them with variable values from a lookup.

https://graphql.org/learn/queries/#variables

https://graphql.org/learn/serving-over-http/#post-request

File: tests/test_product.py (hunk 4 above)
Comment (Phoeneix) [reply tests/test_product.py:473]: Yes? There are a new line at the end of every file as it required by syntax verifiers.

File: tests/test_rasterdefinition.py (hunk 5 above)
Comment (joshgarde) [reply tests/test_rasterdefinition.py:418]: Isn't CapWords/CamelCase utilized for class naming? These are private methods. Also, the `_` prefix convention seems like it would fit here:

> In addition, the following special forms using leading or trailing underscores are recognized (these can generally be combined with any case convention):
> 
>   `_single_leading_underscore`: weak “internal use” indicator. E.g. from M import * does not import objects whose names start with an underscore.

File: tests/test_product.py (hunk 4 above)
Comment (joshgarde) [reply tests/test_product.py:473]: image

GitHub is showing it as an extra new line

File: API/SWODLR/product.py (hunk 2 above)
Comment (Phoeneix) [reply API/SWODLR/product.py]: In a way, but graphQL can't handle empty arguments.
If I don't add those if lines, then I will get an error message.
- products {-> accepted
- products () {-> error message
- products (limit: ) {-> error message
- products (limit: 10) {-> accepted
- products (limit: 10, after: "") {-> error message
- products (limit: 10, after: "{a valid product Id}") {-> accepted

Edit:
So yeah, those variables are nice, but they only work if everything is matching. If it's empty because the input doesn't need it, then you get an error that you forget to correctly file it. That's the if cases for, to only add the variables when they are used, else they will just cause an error...

Comment (joshgarde) [reply API/SWODLR/product.py]: Instead of leaving the values out or setting it to an empty string (`""`), does it work when those values are set to `null`?

Comment (Phoeneix) [reply API/SWODLR/product.py]: products(limit: 50, after: "") or products(limit: 50, after: None) -> 
{"errors":[{"message":"Invalid Syntax : There are more tokens in the query that have not been consumed offending token '}' at line 27 column 13","locations":[{"line":27,"column":13}],"extensions":{"classification":"InvalidSyntax"}}]}

Edit:
So if after is not used, it's easier to not add it at all... That's why the code is full with ifs, to handle this behavior of graphql.

File: tests/test_product.py (hunk 4 above)
Comment (Phoeneix) [reply tests/test_product.py:473]: There is only one new line at the end of the file as intended. Like any other files. I don't see any unintended extra now line.

File: API/SWODLR/product.py (hunk 2 above)
Comment (joshgarde) [reply API/SWODLR/product.py]:
```
products(limit: 50, after: "") 
products(limit: 50, after: None)
```

`None` isn't valid null syntax in GraphQL; can you try `null` instead within a variable map?

File: tests/test_rasterdefinition.py (hunk 5 above)
Comment (Phoeneix) [reply tests/test_rasterdefinition.py:418]: ok, Added _ before the method names, but still using the CapWords to signal it's a method not a variable

File: API/SWODLR/product.py (hunk 2 above)
Comment (Phoeneix) [reply API/SWODLR/product.py]: Updated graplQl message generation, to remove some if statement

File: API/SWODLR/product.py [hunk 6]
```diff
@@ -23,7 +25,7 @@ def GetProductsOfCurrentUser(
         graphQlBody = '''
             {
             	currentUser {
-            		products {
+            		products(limit: ''' + f'{pageSize}, after: "{pageAfterId}"' + ''') {
```
Comment (joshgarde) [inline API/SWODLR/product.py:28]: This seems like a perfect place for using GraphQL variables

https://graphql.org/learn/queries/#variables

File: API/SWODLR/rasterdefinition.py [hunk 7]
```diff
@@ -12,18 +12,31 @@
...
             	currentUser {
-            		rasterDefinitions {
+            		rasterDefinitions(
+                        id:''' + f'"{id}", \
+                        outputGranuleExtentFlag: {outputGranuleExtentFlagString}, \
+                        outputSamplingGridType: {outputSamplingGridType}, \
+                        rasterResolution: {rasterResolution}, \
+                        utmZoneAdjust: {utmZoneAdjust}, \
+                        mgrsBandAdjust: {mgrsBandAdjust}' + ''')
+                    {
```
Comment (joshgarde) [inline API/SWODLR/rasterdefinition.py:39]: Migrate to GraphQL variables

File: API/SWODLR/status.py [hunk 8]
```diff
@@ -33,7 +33,8 @@ def GetStatusByProductID(
                 limitString = ErrorInjector.ValueTypeInjector(expectedType, limitString)
         graphQlBody = '''
             {
-                statusByProduct(product: ''' + productIdString + ''', limit: ''' + limitString + ''') {
+                statusByProduct(product: ''' + productIdString + ''', limit: ''' + limitString
+        graphQlBody += ''') {
```
Comment (joshgarde) [inline API/SWODLR/status.py:37]: Migrate to GraphQL variables

File: tests/test_rasterdefinition.py (hunk 5 above)
Comment (joshgarde) [reply tests/test_rasterdefinition.py:418]: Can these be switched to snake case?

**Review by joshgarde — CHANGES_REQUESTED** [review]
Good work - just a few changes requested

File: API/SWODLR/product.py (hunk 6 above)
Comment (Phoeneix) [reply API/SWODLR/product.py:28]: I am using variables
I am open to suggestions what you mean by GraphQL variables and how do they look in Python code like in this case.

File: API/SWODLR/rasterdefinition.py (hunk 7 above)
Comment (Phoeneix) [reply API/SWODLR/rasterdefinition.py:39]: Those are variables for the GraphQL request, I am not aware if there are different type of variables in Python

File: API/SWODLR/product.py (hunk 6 above)
Comment (joshgarde) [reply API/SWODLR/product.py:28]: GraphQL variables mean utilizing placeholder values `$limit` and then stating the values for those placeholder values within the `variables` key in the request.

https://graphql.org/learn/queries/#variables

https://graphql.org/learn/serving-over-http/#post-request

These are some references for that behavior. For example, a request that utilizes GraphQL variables will look like this:
```
{
  "query": "query($limit: Int, $after: String) { currentUser { products(limit: $limit, after: $after) { id } } }",
  "variables: {
    "limit": 10,
    "after": null
  }
}
```

Comment (Phoeneix) [reply API/SWODLR/product.py:28]: And how will it make the code more readable or easier to understand? It just makes it more confusing as you need to jump around to understand what happens. It does not help for someone with low level of knowledge in GraphQL.

This line is straightforward. You have these input arguments and you put the value after them. Period, clear.
`products(limit: ''' + f'{pageSize}, after: "{pageAfterId}"' + ''') {`

In what you suggested you define another layer of variables. Putting the Python variable into another variable and than using that. Why do we need to define another 'limit' variable as Integer, when we already had that in python? It feels excessive and needlessly complicating the code.

In my opinion the one without that forced new variable layer are much more clean and easier to read. If we would use the limit multiple places in that call, it would make sens to make a variables so we need to update it at only one place, but for one argument adding that much extra code just makes it looks more complicated, harder to understand and creates maintenance overhead.
`{
  "query": "query($limit: Int, $after: String) { currentUser { products(limit: $limit, after: $after) { id } } }",
  "variables: {
    "limit": 10,
    "after": null
  }
}`
vs
`{
  "query": "{ currentUser { products(limit: 10, after: null) { id } } }",
}`

Having those variables does not decrease any logic from the python side, so it does not have any positive effect on the code.

File: tests/test_rasterdefinition.py (hunk 5 above)
Comment (Phoeneix) [reply tests/test_rasterdefinition.py:418]: Why change your mind on naming conventions? Should I redesign all the methods for Snake case then? It would look unprofessional if some methods are in that format and some are not.
What advantage Snake_Case have over CapWords/CamelCase to make the need to redesign the whole naming convention for the repository?

Comment (joshgarde) [reply tests/test_rasterdefinition.py:418]: It looks like the other methods follow a hybrid snakecase/camel/capwords; I'm only asking that the private methods follow snakecase so it falls in line with PEP8. Other methods do not need to be changed

Example of hybrid casing:
```test_Rasterdefinition_GetRasterDefinitionsForCurrentUser_Filter_Resolution_500```

File: API/SWODLR/product.py (hunk 6 above)
Comment (joshgarde) [reply API/SWODLR/product.py:28]: GraphQL variables remove the need to escape the input on the query side. It's the same logic that is utilized for SQL prepared statements - decoupling the query statement from the user input; removing the need to manually perform the string escaping/type casting logic.

The UI for example, utilizes GraphQL variables in its operation and it would be preferable to have these tests be unified with the way we expect users to use the API. This request was also made yesterday: https://github.com/podaac/swodlr-functional-tests/pull/5#discussion_r1488677045

Comment (Phoeneix) [reply API/SWODLR/product.py:28]: > GraphQL variables remove the need to escape the input on the query side.

You to still need to put the input into the same json, just a different part of it, which will look the same just adding a lot of variable related code that will make the query much more confusing.

> decoupling the query statement from the user input; removing the need to manually perform the string escaping/type casting logic.

It does not remove the need to do the logic. I still need to cast Boolean in Python to string as GraphQL can't handle 'False', but only 'false', which is not a valid parameter for Boolean in Python. Same with 'null', Python has 'None', and GraphQL unable to handle that. So no, adding that overhead won't remove any type casting logic/code on Python side.

So just to be clear, you want me to write overcomplicated code that has no practical usage over the simple one in the case of these API tests just because the GraphQL is capable of that? And you expect that everyone will use the complicated version instead of the simple one?

Comment (frankinspace) [reply API/SWODLR/product.py:28]: I wouldn't go so far to say it's impractical; the graphql docs explain it is considered a 'best practice' to separate the dynamic components of the query from the static parts:

> It wouldn't be a good idea to pass these dynamic arguments directly in the query string, because then our client-side code would need to dynamically manipulate the query string at runtime, and serialize it into a GraphQL-specific format. Instead, GraphQL has a first-class way to factor dynamic values out of the query, and pass them as a separate dictionary. ... This is also in general a good practice for denoting which arguments in our query are expected to be dynamic - we should never be doing string interpolation to construct queries from user-supplied values.

Looking at how you've structured the code you would likely have to change your `GraphQlBase.Base_Post` function to accept an operation string and an optional variable dictionary

Comment (Phoeneix) [reply API/SWODLR/product.py:28]: > It wouldn't be a good idea to pass these dynamic arguments directly in the query string, because then our client-side code would need to dynamically manipulate the query string at runtime, and serialize it into a GraphQL-specific format

So they mean they didn't wanted to implement type check on their code side, and having the variables force it is they work around. That's nice and workable for languages where variable types can be very dynamic, so they are recommending this way for type precheck.
I feel that best practice is intended for web pages and things like postman, where the user can randomly add any input without anything forcing type before.
That's why I asked why should I add an extra variable layer and develop new logic to work around it to verify when I forward an integer to the code, there is really an integer, when python already made sure of it?
There are no documentation of what are our input variables and what is the input type to accommodate this best practice and I am not planning to deep dive code to figure that out. I don't think anyone would be comfortable creating the API calls in that way.
I needed to hunt for information for weeks just to make this basic version of it work.
In recommendation if those are the calls you are intending other's to make, then a documentation would be very helpful where all the input variables for each call is listed and described by what are the their input type, what they do, are they mandatory or not and what are the default ignore parameter, like "null". With examples for each how the call should look and not just a reference to check out the code and try to understand and untangle the information from there.

**Comment by Phoeneix** [comment]
I like to point out that everyone seems to forget.
This repository was made to test the SWODLR API behavior, which covers mostly the edge cases and what can go wrong.
The majority of the tests aimed for those situations, not to cover the recommended best practices, which are there to try to cover up most of the errors that the user can do with the basic GraphQL call.
As a note, it that does not make the user stop being able to do them if they do not follow the best practice recommendations.

File: API/SWODLR/product.py (hunk 6 above)
Comment (joshgarde) [reply API/SWODLR/product.py:28]: > It does not remove the need to do the logic. I still need to cast Boolean in Python to string as GraphQL can't handle 'False', but only 'false', which is not a valid parameter for Boolean in Python. Same with 'null', Python has 'None', and GraphQL unable to handle that. So no, adding that overhead won't remove any type casting logic/code on Python side.

By utilizing GraphQL variables, it removes the need to typecast any of the variables datatypes on the query side. That's because the variables lookup is represented as a JSON object which Python can natively convert dict objects to. It's as simple as `json.dumps({'boolVal': False, 'stringVal': 'string', 'intVal': 0})`.

> There are no documentation of what are our input variables and what is the input type to accommodate this best practice and I am not planning to deep dive code to figure that out. I don't think anyone would be comfortable creating the API calls in that way.

GraphQL was designed to accommodate this type of design and we have documentation on the input types of every call in SWODLR's API build in as part of the API specification.

https://github.com/podaac/swodlr-api/tree/develop/src/main/resources/graphql

> I needed to hunt for information for weeks just to make this basic version of it work.

The input types are already defined as part of the API specification. The main work required here would be to parameterize the queries and to introduce a mechanism to input a variables dict to `GraphQlBase.Base_Post` like Frank suggested.

**Comment by joshgarde** [comment]
> This repository was made to test the SWODLR API behavior, which covers mostly the edge cases and what can go wrong.
The majority of the tests aimed for those situations, not to cover the recommended best practices, which are there to try to cover up most of the errors that the user can do with the basic GraphQL call.

I agree that we are trying to test edge cases/API call issues, however, we aren't trying to test our GraphQL library's syntax parsing implementation. We're limiting scope specifically to SWODLR's API behavior which is higher level than the GraphQL syntax errors that a new API user might run into. I think that by utilizing GraphQL variables, we're not only making the testing more readable and expandable in the future, but we're also avoiding common syntax/typecasting pitfalls that we're specifically not testing for. Utilizing GraphQL variables means we can avoid needing to typecast within the tests and jump straight into testing core functionality instead of GraphQL syntax parsing

File: API/SWODLR/product.py (hunk 6 above)
Comment (Phoeneix) [reply API/SWODLR/product.py:28]: > https://github.com/podaac/swodlr-api/tree/develop/src/main/resources/graphql

That is not a documentation, that is some code without any explanation... It does not make me understand what it should do, what were the mandatory and optional parameters, expected inputs, input ranges and so. I needed to ask you again and again for information.
Let me give you some example of API documentation:
https://nasa.github.io/cumulus-api/v18.1.0/#cumulus-api
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ecs/client/list_services.html
https://support.testrail.com/hc/en-us/articles/7077723976084-Milestones#getmilestones

> The input types are already defined as part of the API specification. The main work required here would be to parameterize the queries and to introduce a mechanism to input a variables dict to GraphQlBase.Base_Post like Frank suggested.

And redesigning 70% of the tests as they are testing negative behavior that the "best practice" will mask or invalidate. As I mentioned before, I am not designing tests that will test best behavior, I am designing test that do edge cases and error handling, which is definitely not the domain of "best  practice".

**Comment by Phoeneix** [comment]
> we're not only making the testing more readable and expandable in the future

We are NOT making the test more readable for anyone except GraphQL experts, who live on that stuff...
A Json is understandable and readable. Putting variables, methods and stuff in the json code make it more confusing and less readable.
Unless there is a practical reason that won't affect negatively the already existing tests, this "best practice" major refactor and redesign work with no reward and added extra complexity is definitely not part of this PR.
Unless there is a new issue made to switch to graphql code that using graphql variables and redesign all the API tests to accommodate that behavior.

**Comment by frankinspace** [comment]
The content of the requests you're sending is GraphQL, not JSON (although it looks similar to JSON). Using variables in queries is one of the basic features of GraphQL; similar to using parameterized queries in SQL.

That said, it isn't a requirement. I agree variables would make the code more maintainable long term. We really only need at most 1 test where we try to submit a `string` for a variable that the schema specifies as an `int`.

**Comment by viviant100** [comment]
This PR includes the following 2 tests:
- Pagination tests
- Filtering tests for RasterDefinitions
